### PR TITLE
[WIP] Add SRV and TXT record lookup methods to the iomgr API

### PR DESCRIPTION
**Pull request by drfloob**
No review needed at this stage. Creating this draft PR for visibility. This will likely land with an ares wrapper refactoring.

### CODE REVIEW

**Comment by drfloob** [comment]
@markdroth this is ready for review.

File: src/core/ext/filters/client_channel/resolver/dns/c_ares/dns_resolver_ares.cc [hunk 1]
```diff
@@ -128,17 +154,32 @@ class AresClientChannelDNSResolver : public PollingResolver {
...
     RefCountedPtr<AresClientChannelDNSResolver> resolver_;
-    std::unique_ptr<grpc_ares_request> request_;
-    grpc_closure on_resolved_;
+    grpc_closure on_hostname_resolved_;
+    std::unique_ptr<grpc_ares_request> hostname_request_;
+    grpc_closure on_srv_resolved_;
+    std::unique_ptr<grpc_ares_request> srv_request_;
+    grpc_closure on_txt_resolved_;
+    std::unique_ptr<grpc_ares_request> txt_request_;
+    Mutex on_resolved_mu_;
```
Comment (markdroth) [inline src/core/ext/filters/client_channel/resolver/dns/c_ares/dns_resolver_ares.cc:182]: Please use lock annotations to indicate what fields are guarded by this mutex.

File: src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.cc [hunk 2]
```diff
@@ -1079,26 +1053,135 @@ static grpc_ares_request* grpc_dns_lookup_ares_impl(
...
+  grpc_core::MutexLock lock(&r->mu);
+  r->ev_driver = nullptr;
+  r->on_done = on_done;
+  r->balancer_addresses_out = balancer_addresses;
+  GRPC_CARES_TRACE_LOG(
+      "request:%p c-ares grpc_dns_lookup_srv_ares_impl name=%s", r, name);
+  grpc_error_handle error = GRPC_ERROR_NONE;
+  // Don't query for SRV records if the target is "localhost"
+  if (target_matches_localhost(name)) {
+    error = grpc_error_set_str(GRPC_ERROR_CREATE_FROM_STATIC_STRING(
```
Comment (markdroth) [inline src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.cc:1100]: This doesn't look like the right behavior. If someone does a lookup on localhost with SRV lookups enabled, we don't want to fail the query; we just want to not return any SRV records.

File: src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.cc [hunk 3]
```diff
@@ -1079,26 +1053,135 @@ static grpc_ares_request* grpc_dns_lookup_ares_impl(
...
+  r->on_done = on_done;
+  r->service_config_json_out = service_config_json;
+  GRPC_CARES_TRACE_LOG(
+      "request:%p c-ares grpc_dns_lookup_txt_ares_impl name=%s", r, name);
+  grpc_error_handle error = GRPC_ERROR_NONE;
+  // Don't query for TXT records if the target is "localhost"
   if (target_matches_localhost(name)) {
-    r->balancer_addresses_out = nullptr;
-    r->service_config_json_out = nullptr;
+    error = grpc_error_set_str(GRPC_ERROR_CREATE_FROM_STATIC_STRING(
```
Comment (markdroth) [inline src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.cc:1141]: Same comment as for SRV lookups above: If someone does a lookup of localhost with TXT records enabled, we should not fail the query, just return no records.

File: src/core/ext/filters/client_channel/resolver/dns/c_ares/dns_resolver_ares.cc [hunk 4]
```diff
@@ -276,15 +317,51 @@ std::string ChooseServiceConfig(char* service_config_choice_json,
...
 }
 
-void AresClientChannelDNSResolver::AresRequestWrapper::OnResolved(
-    grpc_error_handle error) {
-  GRPC_CARES_TRACE_LOG("resolver:%p OnResolved()", this);
+void AresClientChannelDNSResolver::AresRequestWrapper::OnResolvedLocked(
+    grpc_error_handle error) ABSL_EXCLUSIVE_LOCKS_REQUIRED(on_resolved_mu_) {
+  if (hostname_request_ != nullptr || srv_request_ != nullptr ||
+      txt_request_ != nullptr) {
+    GRPC_CARES_TRACE_LOG(
```
Comment (markdroth) [inline src/core/ext/filters/client_channel/resolver/dns/c_ares/dns_resolver_ares.cc:357]: Should this log message be outside of this block?  As currently written, it will be logged only if all three requests are null, so the "waiting" messages will never be logged.

File: src/core/ext/filters/client_channel/resolver/dns/c_ares/dns_resolver_ares.cc [hunk 5]
```diff
@@ -276,15 +317,51 @@ std::string ChooseServiceConfig(char* service_config_choice_json,
...
+        "resolver:%p OnResolved() waiting for results (hostname: %s, srv: %s, "
+        "txt: %s)",
+        this, hostname_request_ ? "waiting" : "done",
```
Comment (markdroth) [inline src/core/ext/filters/client_channel/resolver/dns/c_ares/dns_resolver_ares.cc:360]: Please use `!= nullptr` for pointer comparisons rather than evaluating them as bools.

File: src/core/ext/filters/client_channel/resolver/dns/c_ares/dns_resolver_ares.cc [hunk 6]
```diff
@@ -416,65 +486,224 @@ class AresDNSResolver : public DNSResolver {
...
+          resolver_(resolver),
+          aba_token_(aba_token) {
+      GRPC_CLOSURE_INIT(&on_dns_lookup_done_, OnDnsLookupDone, this,
+                        grpc_schedule_on_exec_ctx);
+      grpc_pollset_set_add_pollset_set(pollset_set_, interested_parties_);
+    }
+
+    // closure to call when the ares resolution request completes. Subclasses
+    // should use this as the ares callback in MakeRequestLocked()
+    grpc_closure on_dns_lookup_done_ ABSL_GUARDED_BY(mu_);
```
Comment (markdroth) [inline src/core/ext/filters/client_channel/resolver/dns/c_ares/dns_resolver_ares.cc:504]: Data members must be private, not protected, as per the style guide.

https://google.github.io/styleguide/cppguide.html#Access_Control

However, you can add protected accessor methods for them.

File: src/core/ext/filters/client_channel/resolver/dns/c_ares/dns_resolver_ares.cc [hunk 7]
```diff
@@ -363,51 +439,45 @@ class AresClientChannelDNSResolverFactory : public ResolverFactory {
 
 class AresDNSResolver : public DNSResolver {
  public:
+  // Abstract class that centralizes common request handling logic via the
```
Comment (markdroth) [inline src/core/ext/filters/client_channel/resolver/dns/c_ares/dns_resolver_ares.cc:442]: The interactions between the base class and the subclasses here is fairly complex and hard to follow.  I think this is a symptom of the problem I pointed out in my other comment, where the paradigm of the current c-ares code does not match that of the EE API.  I think if we clean up that underlying problem, then this code will get a lot simpler, because a lot of the state that is currently being stored in the `AresRequest` base class will instead be able to be stored in the `DNSResolver` class itself, which all of the individual request objects will have a ref to.

File: src/core/ext/filters/client_channel/resolver/dns/c_ares/dns_resolver_ares.cc [hunk 8]
```diff
@@ -416,65 +486,224 @@ class AresDNSResolver : public DNSResolver {
...
+        resolved_addresses.reserve(addresses_->size());
+        for (const auto& server_address : *addresses_) {
+          resolved_addresses.push_back(server_address.address());
+        }
+      }
+      on_resolve_address_done_(std::move(resolved_addresses));
+    }
+
     // the name to resolve
     const std::string name_;
```
Comment (markdroth) [inline src/core/ext/filters/client_channel/resolver/dns/c_ares/dns_resolver_ares.cc:591]: The `name_`, `name_server_`, and `timeout_` members are common to all subclasses.  Maybe move these to the base class, with protected accessor methods?

**Review by markdroth — CHANGES_REQUESTED** [review]
This looks like a good start!

Please let me know if you have any questions.  Thanks!

File: src/core/lib/iomgr/resolve_address.h [hunk 9]
```diff
@@ -60,17 +61,38 @@ class DNSResolver {
...
 
   // Resolve name in a blocking fashion. Use \a default_port if a port isn't
   // designated in \a name, otherwise use the port in \a name.
   virtual absl::StatusOr<std::vector<grpc_resolved_address>>
-  ResolveNameBlocking(absl::string_view name,
-                      absl::string_view default_port) = 0;
+  LookupHostnameBlocking(absl::string_view name,
+                         absl::string_view default_port) = 0;
+
+  // Asynchronously resolve an SRV Record per gRFC A5: Load Balancing and DNS.
```
[the remark this answers is not in the document]
Comment (drfloob) [reply src/core/lib/iomgr/resolve_address.h:76]: Good point, removed.

File: src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.cc [hunk 10]
```diff
@@ -827,46 +828,23 @@ static void on_txt_done_locked(void* arg, int status, int /*timeouts*/,
...
-  }
-  error = grpc_ares_ev_driver_create_locked(&r->ev_driver, interested_parties,
-                                            query_timeout_ms, r);
-  if (!GRPC_ERROR_IS_NONE(error)) goto error_cleanup;
-  // If dns_server is specified, use it.
-  if (dns_server != nullptr && dns_server[0] != '\0') {
-    GRPC_CARES_TRACE_LOG("request:%p Using DNS server %s", r, dns_server);
+grpc_error_handle set_request_dns_server(grpc_ares_request* r,
+                                         absl::string_view dns_server,
+                                         const char* name)
```
[the remark this answers is not in the document]
Comment (drfloob) [reply src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.cc:833]: SGTM, done.

File: src/core/ext/filters/client_channel/resolver/dns/c_ares/dns_resolver_ares.cc (hunk 8 above)
Comment (drfloob) [reply src/core/ext/filters/client_channel/resolver/dns/c_ares/dns_resolver_ares.cc]: Good call, done.

File: src/core/lib/iomgr/resolve_address.h [hunk 11]
```diff
@@ -60,17 +61,38 @@ class DNSResolver {
...
+  // On completion, \a on_done is invoked with the result.
+  //
+  // The same caveats in \a LookupHostname apply here as well.
+  //
+  // Reference: https://github.com/grpc/proposal/blob/master/A5-grpclb-in-dns.md
+  virtual TaskHandle LookupSRV(
+      std::function<void(absl::StatusOr<std::vector<grpc_resolved_address>>)>
```
[the remark this answers is not in the document]
Comment (drfloob) [reply src/core/lib/iomgr/resolve_address.h:81]: That is the plan, but currently that's a step away. I've already made some progress on it in a followup branch that brings in other EE changes as well. I can consider merging that in to this PR instead, but felt it belonged with a broader set of EE type and semantics changes.

File: src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.h [hunk 12]
```diff
@@ -78,23 +78,36 @@ struct grpc_ares_request {
...
-extern grpc_ares_request* (*grpc_dns_lookup_ares)(
+extern grpc_ares_request* (*grpc_dns_lookup_hostname_ares)(
     const char* dns_server, const char* name, const char* default_port,
     grpc_pollset_set* interested_parties, grpc_closure* on_done,
     std::unique_ptr<grpc_core::ServerAddressList>* addresses,
+    int query_timeout_ms);
+
+// Asynchronously resolve a SRV record.
+// See \a grpc_dns_lookup_hostname_ares for usage details and caveats.
+extern grpc_ares_request* (*grpc_dns_lookup_srv_ares)(
```
[the remark this answers is not in the document]
Comment (drfloob) [reply src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.h:100]: Agreed with the concern here, thanks for bringing it up. I went into this expecting that 3x the fds & 3x the memory are negligible given that most users will invoke a small number of short-lived, concurrent requests at any given time. I'd be interested to hear if that isn't the case.

> once we've already finished migrating the code into the EE impl, there won't be much motivation to do this refactoring

This is a common thing indeed, but since I'm doing the development on this at the moment, I don't think it's a huge risk. I subscribe to the "first: make it work; second: make it better" philosophy. Given that Alex already paved the way for an improved refactoring, I'd opt to land this change that lets us leverage the new API, and do the refactoring if we find it would be measurably meaningful (this is a flexible stance on your (1) or (3) distinction, we can adapt as needed).

That said, @apolcyn was it terribly difficult to debug the problems with the refactored resolver? I'm curious about your time investment there.

File: src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.cc (hunk 3 above)
Comment (drfloob) [reply src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.cc]: Thanks, fixed.

File: src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.cc (hunk 2 above)
Comment (drfloob) [reply src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.cc]: Thanks, fixed.

File: src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.h (hunk 12 above)
[the remark this answers is not in the document]
Comment (apolcyn) [reply src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.h:100]: I think I favor option 1), because this should amount to the simplest code, and at the moment I'm skeptical that the benefits of the shared state approach are worth the complexity.

In particular, I'm thinking about the following:

a) Since grpclb is deprecated, SRV queries should actually go away long term.

b) TXT queries aren't enabled by default. This point might be reconsidered once they are or are soon to be.

c) I don't know of a use case that really relies on large amounts of concurrent SRV and/or TXT queries. E.g.., personally I'd be surprised if an extra file descriptor or two per DNS resolution caused problems.

> That said, @apolcyn was it terribly difficult to debug the problems with the refactored resolver? I'm curious about your time investment there.

The problem I ran into was basically a problem with an integration test I was unfamiliar with, and which I couldn't (easily) run myself or gather more info about, to debug further. I then just didn't find time to push forward on it :(

That said, when I originally submitted that PR, it was using a work serializer. I know at least one bug fix has been made to the work serializer since then. Also, the c-ares wrapper code has since changed to use a mutex. So I wouldn't be surprised if the bug has been fixed in the interim, and maybe we could retry the test.

File: src/core/ext/filters/client_channel/resolver/dns/c_ares/dns_resolver_ares.cc (hunk 5 above)
Comment (drfloob) [reply src/core/ext/filters/client_channel/resolver/dns/c_ares/dns_resolver_ares.cc]: Regardless, it should be consistent throughout the PR, which it wasn't. Fixed.

File: src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.h (hunk 12 above)
[the remark this answers is not in the document]
Comment (markdroth) [reply src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.h:100]: As we discussed, the thing I'm concerned about here is not your individual commitment to follow up on the clean-up here, but rather whether we as a team will wind up giving you the time to do that.  What usually seems to happen to all of us is that once things are working, we get pushed toward subsequent priorities, so if we don't take time to get things right before we finish them, we never seem to actually get time to do the cleanup.

How about this: Let's go forward with this PR as-is for now, but let's fix this as part of moving the code from iomgr to the EE impl, rather than doing it as a separate clean-up step later.  Does that work?

File: src/core/lib/iomgr/resolve_address.h (hunk 11 above)
[the remark this answers is not in the document]
Comment (drfloob) [reply src/core/lib/iomgr/resolve_address.h:84]: Done.

File: src/core/lib/iomgr/resolve_address.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (drfloob) [reply src/core/lib/iomgr/resolve_address.h]: Oops, fixed. Thanks.

File: src/core/ext/filters/client_channel/resolver/dns/c_ares/dns_resolver_ares.cc [hunk 13]
```diff
@@ -153,7 +153,7 @@ class AresClientChannelDNSResolver : public PollingResolver {
       resolver_.reset(DEBUG_LOCATION, "dns-resolving");
     }
 
-    void Orphan() override {
+    void Orphan() override ABSL_NO_THREAD_SAFETY_ANALYSIS {
```
[the remark this answers is not in the document]
Comment (drfloob) [reply src/core/ext/filters/client_channel/resolver/dns/c_ares/dns_resolver_ares.cc:159]: Done

File: src/core/ext/filters/client_channel/resolver/dns/c_ares/dns_resolver_ares.cc [hunk 14]
```diff
@@ -487,25 +494,27 @@ class AresDNSResolver : public DNSResolver {
     }
 
    protected:
-    AresRequest(grpc_pollset_set* interested_parties, AresDNSResolver* resolver,
-                intptr_t aba_token)
-        : pollset_set_(grpc_pollset_set_create()),
+    AresRequest(absl::string_view name, absl::string_view name_server,
+                Duration timeout, grpc_pollset_set* interested_parties,
+                AresDNSResolver* resolver, intptr_t aba_token)
+        : name_(std::string(name)),
```
[the remark this answers is not in the document]
Comment (drfloob) [reply src/core/ext/filters/client_channel/resolver/dns/c_ares/dns_resolver_ares.cc]: Done.

File: src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.h (hunk 12 above)
[the remark this answers is not in the document]
Comment (apolcyn) [reply src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.h:100]: @markdroth by cleanup, do we mean finishing the C++ification, or by sharing the event driver between the different lookups?

If the latter, personally I'm skeptical if that's a net win, for those reasons above.

File: src/core/ext/filters/client_channel/resolver/dns/c_ares/dns_resolver_ares.cc (hunk 7 above)
Comment (drfloob) [reply src/core/ext/filters/client_channel/resolver/dns/c_ares/dns_resolver_ares.cc:452]: See other thread for agreed-upon terms of a "cleanup" and ownership stories.

File: src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.h (hunk 12 above)
[the remark this answers is not in the document]
Comment (drfloob) [reply src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.h:100]: @apolcyn I welcome you to pick up the C++ification again if you want. Otherwise, I don't think "cleanup" is necessarily the right term here, it's just a different, valid implementation. I'm still in favor of taking the flexible stance of landing this and paying attention to those who may be the most sensitive customers to determine if there's any meaningful change in performance that would necessitate a more efficient implementation.

[the remark this answers is not in the document]
Comment (markdroth) [reply src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.h:100]: If nothing else, I think there will be additional memory usage here from allocating all of the state separately for each request, and given how expensive RAM is right now, that doesn't seem like a win.  Prior to this PR, we were not paying that cost, so this is making things worse, and I don't think we should do that if we have a reasonable alternative -- and I think we do.  I am willing to live with this in the short term, but I want to make sure that we will clean it up before we are finished with the EE migration.

As I said to AJ offline, I think that we are going to need to make changes to the event-loop integration anyway as part of moving this code into the EE impl, since that integration will no longer use things like pollset_set.  And since we're going to be changing that code anyway, I think centralizing it into a parent object should be reasonable to do at the same time.

As a completely separate matter, I think that if we are making a significant changes to a piece of code anyway, we should convert that code to C++ if it's still in C.  That does not necessarily mean that we need to convert every line of c-ares code, but if we're touching some given part of it anyway, I think it makes sense to convert it.  I would say the same thing about any piece of code in our codebase.

Alex, I think we should also try to learn from this for the future.  If we don't get your PR rolled forward -- and neither of you seem to want to take the time to do that, so it seems like it's not going to happen -- then the fact that we didn't follow up on it means that all the time you spent on it was basically completely wasted, because we didn't wind up getting any benefit from it.  I think we should try really hard to avoid that in the future.  If something is worth spending time on, it's worth doing it right and getting it done; if it's not, then we shouldn't spend time on it in the first place.

**Review by markdroth — APPROVED** [review]
